### PR TITLE
Check ctrlp inside loop

### PR DESCRIPTION
**Pull request by vext01**
Main change is in the first commit.

https://github.com/ykjit/yk/pull/475 includes a test for this.

The compiler-side should be merged first.

### CODE REVIEW

File: llvm/lib/Transforms/Yk/ControlPoint.cpp [hunk 1]
```diff
@@ -121,8 +122,24 @@ class YkControlPoint : public ModulePass {
     // Get function containing the control point.
     Function *Caller = OldCtrlPointCall->getFunction();
 
-    // Find all live variables just before the call to the control point.
+    // Check that the control point was called inside, what appears to be, an
+    // interpreter loop, i.e. in a top-level loop in some function.
     DominatorTree DT(*Caller);
+    const LoopInfo Loops(DT);
+    bool FoundInLoop = false;
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/ControlPoint.cpp:129]: Is this equivalent to something like `Loops.any_of(..)` (which I think might be the C++ equivalent of Rust's `any` in an iterator)?

File: llvm/lib/Transforms/Yk/ControlPoint.cpp [hunk 2]
```diff
@@ -121,8 +122,24 @@ class YkControlPoint : public ModulePass {
     // Get function containing the control point.
     Function *Caller = OldCtrlPointCall->getFunction();
 
-    // Find all live variables just before the call to the control point.
+    // Check that the control point was called inside, what appears to be, an
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: This comment is a bit confusing: can we not just say "Check that the control point is inside a loop." Also, does the code below actually check only top-level loops? Do we even need that as a constraint? I guess it doesn't *hurt* to have it as a constraint though.

Comment (vext01) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: > can we not just say "Check that the control point is inside a loop." 

Not without a more complex check, as we'd have to recursively check sub-loops (LLVM's `LoopInfo`is a tree of loops).

I considered searching deeper, but ended up concluding that it's not really worth the hassle for now.

What do you think?

Comment (ltratt) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: Ah, so this really does check only top-level loops? Fine! Then let's make the comment "Check that the control point is in a top-level loop." without the extra text?

Comment (ltratt) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: Yes please!

Comment (vext01) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: It appears that the iterator over the loops is already recursing into sub-loops, so if you are OK with it, we can just say "yk_control_point() must be called inside a loop".

What do you think?

Comment (vext01) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: bd256da implements that.

File: llvm/lib/Transforms/Yk/ControlPoint.cpp (hunk 1 above)
Comment (vext01) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp]: Fixed in bd256da

File: llvm/lib/Transforms/Yk/ControlPoint.cpp (hunk 2 above)
Comment (ltratt) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: Oh, wow, I've made you made it worse -- sorry! I think we should go back with what you had. I'm a bit stunned there's not a nicer way of doing it!

Comment (vext01) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: Just to check, are we talking about the `any_of` here, and not whether we search inner loops or not?

The `any_of` is verbose, but actually fairly idiomatic of C++. A seasoned C++ programmer would still prefer what you proposed, I think.

Comment (ltratt) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:125]: Ah, OK, then let's keep it then!

**Comment by ltratt** [comment]
We need to fix the comment but apart from that I think we're good to go!

**Comment by vext01** [comment]
> 

bd256da adjusted the comment (to not mention top-level loops). Is that OK, or does it need further adjustment?

 [We do need to adjust the commit message at squash time]

**Comment by ltratt** [comment]
The comment is still unnecessarily complex, and not in a useful way, IMHO: I think it would be better to simplify it more along the lines of my suggestion.

**Comment by vext01** [comment]
Do you mean like in 514467a?

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat

**Pull request merged**